### PR TITLE
Handle lamport clocks

### LINKED ISSUE #16: Add support for lamport clocks for concurrent inserts at the same position

### PR DESCRIPTION
**Pull request by katzper-michno**
Resolves #16

### CODE REVIEW

File: server.py [hunk 1]
```diff
@@ -154,12 +157,16 @@ def handle_document_delta(self, message):
...
-                    idx, = (i for i, val in enumerate(self.document) if val['id'] == position)
-                    self.document[idx + 1: idx + 1] = insert
+                for char in insert:
+                    self.lamportClock = max(self.lamportClock, int(char['clock']))
+                prevId = insert[0]['prevId']
+                clock = insert[0]['clock']
+                idx = 0
+                if prevId != None:
+                    idx, = (i for i, val in enumerate(self.document) if val['id'] == prevId)
+                while idx < len(self.document) and self.document[idx]['prevId'] == prevId and self.document[idx]['clock'] > clock:
```
Comment (przyjacielpkp) [inline server.py]: consider changing this line using  `next(x in self.documents if x['prevId'] == prevId and x['clock'] > clock)` and similarly in the line above

File: server.py [hunk 2]
```diff
@@ -154,12 +157,16 @@ def handle_document_delta(self, message):
...
             insert = message['insert']
             if len(insert) > 0:
-                position = insert[0]['prevId']
-                if position == None:
-                    self.document[0:0] = insert
-                else:
-                    idx, = (i for i, val in enumerate(self.document) if val['id'] == position)
-                    self.document[idx + 1: idx + 1] = insert
+                for char in insert:
+                    self.lamportClock = max(self.lamportClock, int(char['clock']))
```
Comment (przyjacielpkp) [inline server.py]: consider  `self.lamportClock =  max(self.lamportClock, max(int(char['clock']) for char in insert ))` here

**Review by ghost — COMMENTED** [review]
LGTM

**Comment by przyjacielpkp** [comment]
blank

**Review by przyjacielpkp — CHANGES_REQUESTED** [review]
that's quite strange I have to re-request review to myself to be able to request changes for PR ( as i was reviewer and assignee simultaneously)

File: server.py (hunk 1 above)
Comment (katzper-michno) [reply server.py]: won't this give me the first element in self.document which satisfies `x['prevId'] == prevId and x['clock'] > clock`? I want the first element with index `i >= idx` that does not satisfy this

File: server.py (hunk 2 above)
Comment (katzper-michno) [reply server.py]: this is nice, will change

File: server.py (hunk 1 above)
Comment (przyjacielpkp) [reply server.py]: Okey i think the proper condition should look like that `next(x for i, x in enumerate(self.documents) if  i >=idx and (['prevId'] != prevId or x['clock'] <= clock) )` ( especially considering the fact, that we have to negate the while condition)

Comment (katzper-michno) [reply server.py]: done, and it seems to be working

File: server.py (hunk 2 above)
Comment (katzper-michno) [reply server.py]: changed

File: server.py [hunk 3]
```diff
@@ -154,12 +157,17 @@ def handle_document_delta(self, message):
         if 'insert' in message:
             insert = message['insert']
             if len(insert) > 0:
-                position = insert[0]['prevId']
-                if position == None:
-                    self.document[0:0] = insert
-                else:
-                    idx, = (i for i, val in enumerate(self.document) if val['id'] == position)
-                    self.document[idx + 1: idx + 1] = insert
+                for char in insert:
```
Comment (przyjacielpkp) [inline server.py]: there is no need for that `for` loop to be here anymore i believe

File: server.py [hunk 4]
```diff
@@ -154,12 +157,17 @@ def handle_document_delta(self, message):
...
+                    self.lamportClock = max(self.lamportClock, max(int(char['clock']) for char in insert))
+                prevId = insert[0]['prevId']
+                clock = insert[0]['clock']
+                idx = 0
```
Comment (przyjacielpkp) [inline server.py:163]: no need for that variable assignment

**Review by przyjacielpkp — APPROVED** [review]
LGTM, just quick fixes left

File: server.py (hunk 3 above)
Comment (katzper-michno) [reply server.py]: removed

**Pull request merged**